src-tauri/src/lib.rs

```diff
@@ -510,3 +510,75 @@ pub fn get_installed_mods(game_install: GameInstall) -> Result<Vec<NorthstarMod>
 
     Ok(installed_mods)
 }
+
+async fn get_ns_mod_download_url(thunderstore_mod_string: String) -> Result<String, String> {
+
+    // TODO: This will crash the thread if not internet connection exist. `match` should be used instead
+    let index = thermite::api::get_package_index().await.unwrap().to_vec();
+
+    // String replace works but more care should be taken in the future
+    let ts_mod_string_url = thunderstore_mod_string.replace("-", "/");
+
+    for ns_mod in index {
+        if ns_mod.url.contains(&ts_mod_string_url) {
+            dbg!(ns_mod.clone());
+            return Ok(ns_mod.url);
+        }
+    }
+
+    Err("Could not find mod on Thunderstore".to_string())
+}
+
+// Copied from `libtermite` source code and modified
+// Should be replaced with a library call to libthermite in the future
+/// Download and install mod to the specified target.
+pub async fn fc_install_mod(
+    game_install: GameInstall,
+    thunderstore_mod_string: String,
+) -> Result<(), String> {
+    // Get mods and download directories
+    let download_directory = format!(
+        "{}/___flightcore-temp-download-dir/",
+        game_install.game_path
+    );
+    let mods_directory = format!("{}/R2Northstar/mods/", game_install.game_path);
+
+    // Get download URL for the specified mod
+    let download_url = get_ns_mod_download_url(thunderstore_mod_string.clone()).await?;
+
+    // Create download directory
+    match std::fs::create_dir_all(download_directory.clone()) {
+        Ok(()) => (),
+        Err(err) => return Err(err.to_string()),
+    };
+
+    let name = thunderstore_mod_string;
+    let path = format!(
+        "{}/___flightcore-temp-download-dir/{}.zip",
+        game_install.game_path, name
+    );
+
+    // Download the mod
+    let f = match thermite::core::actions::download_file(&download_url, path.clone()).await {
+        Ok(f) => f,
+        Err(e) => return Err(e.to_string()),
+    };
+
+    // Extract the mod to the mods directory
+    let pkg = match thermite::core::actions::install_mod(&f, std::path::Path::new(&mods_directory))
+    {
+        Ok(pkg) => pkg,
+        Err(err) => return Err(err.to_string()),
+    };
+    dbg!(pkg.clone());
+
+    // Delete downloaded zip file
+    std::fs::remove_file(path).unwrap();
+
+    // Delete temp download folder
+    // TODO: For some reason this fails. Maybe delete download folder in separate call?
```

Comment:
Why doesn't this work?

Reply:
Uncommenting it and then installing a mod gives:

```
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 303, kind: Uncategorized, message: "File cannot be opened cause it is currently being deleted." }', E:\Dateien\Coding\FlightCore\src-tauri\src\lib.rs:581:49
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

(translated error message from German by hand)

Reply:
Can't you just remove the `___flightcore-temp-download-dir/` directory?

Reply:
What do you mean exactly?

Delete it after done installing mod? Cause that's what the commented out function is supposed to do.

Don't use the temp folder in the first place? Not a big fan cause a crash mid-install will cause leftover files littering the Titanfall2 directory as opposed to being restricted to a single folder ^^

Reply:
Yeah, I mean only deleting directory after mod installation; I suggest to keep only the `std::fs::remove_dir_all(download_directory).unwrap();` call.

While commenting all remove calls, I sometimes have another exception:
```rust
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: StripPrefixError(())', C:\Users\Remy\.cargo\registry\src\github.com-1ecc6299db9ec823\libthermite-0.2.0\src\core\actions.rs:197:46
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

`The_Peepeepoopoo_man-Chimera-0.1.1` causes this issue (creating a `1667515294` folder with zip content in my `mods` directory), while `sammydoodlew-Volt_Vaporwave-1.1.0` seems to install just fine.

Reply:
> Yeah, I mean only deleting directory after mod installation; I suggest to keep only the `std::fs::remove_dir_all(download_directory).unwrap();` call.

Both directory remove calls don't work individually, I had them both in there still as comments as a reminder to me. Didn't intend to use both :P

> ```rust
> thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: StripPrefixError(())', C:\Users\Remy\.cargo\registry\src\github.com-1ecc6299db9ec823\libthermite-0.2.0\src\core\actions.rs:197:46
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> ```

In regards to that panic is in `libthermite`, might need to check with Emerald on the issue with the lib (although our version of the lib is outdated atm).

Another thing I realised, is that if we do multiple mod installs at once e.g. user clicking on update buttons for multiple mods, deleting the folder in the install call will delete the folder while the second mod install is still going.

So maybe we should just resort to deleting temp download folder in a separate function call. ^^

Reply:
> So maybe we should just resort to deleting temp download folder in a separate function call. ^^

Nah, this method should remove all folders it had created.
A simple solution is not to use same name - for all temporary folders, but rather to use mod name and/or current timestamp.

Reply:
That would work. My main concern is leftover files from failed operations...

So maybe a common prefix (the same one we use rn) that could be used to identify folder/files created by FlightCore, so they can be easily removed again if necessary?

Reply:
> In regards to that panic is in libthermite, might need to check with Emerald on the issue with the lib (although our version of the lib is outdated atm).

Published v0.3.2 this morning, that panic now returns an error and the temp directory `install_mod` creates should now only exist in the scope of the function

Reply:
@GeckoEidechse will updating `libthermite` solve this issue?

Reply:
It should yes. But before updating `libthermite` I'd like to get #38 merged and then make a new release so that I can better track down any errors, should they occur.

Reply:
With `0.3.3` of `libthermite` installing `The_Peepeepoopoo_man-Chimera-0.1.1` now throws a proper error message ^^

![image](https://user-images.githubusercontent.com/40122905/201550868-8fc6955b-02be-4851-a981-1f990c931b92.png)

Reply:
Given that zip file gets removed and all that is left-over is an empty folder, I have decided that this is a non-issue, to be dealt with by future-Gecko™ in a separate PR ¯\\\_(ツ)\_/¯

Reply:
That definitely is an issue, though. :)

Reply:
Done with c74de4fa7bbceb019931bae3244acd55e3114b8c